### ISSUE TITLE
Clean removal of brokers from envoy ingress controller on cluster scale-down

### ISSUE THREAD

**Issue by amuraru** [banzaicloud/kafka-operator#440]
**Is your feature request related to a problem? Please describe.**
While implementing https://github.com/banzaicloud/kafka-operator/issues/438 I realised we can improve the way ingress
controller manages the configuration for upstream kafka broker proxying.

Currently the envoy config is generated exclusively based of `KafkaCluster.Spec.Brokers` field. 
This works in practice for initial setup and scale-up operations as the newly added brokers are added in ingress but won't be "advertised" as partitions owners by kafka internal metadata mechanism until they actually start serving partitions.
So it doesn't really require to synchronise the ingress config with actual kafka brokers being operational. (i.e. having a broker exposed in ingress but not yet returned in Metadata Responses from Kafka is a non-issue)

When scaling down the cluster though, the sequence of removing the brokers from ingress matters.
Here is an example of events:
1. Start with a cluster of say 6 brokers
2. Envoy ingress is configured to proxy all of them
3. Remove a broker from ``KafkaCluster.Spec.Brokers`
kafka and envoy reconcilers are both notified of this change and they async operate the corresponding changes:
(3.1) envoy reconciler is refreshing the envoy.yaml configuration removing the broker
(3.2) kafka reconciler starts draining the broker and waiting for cruise-control to complete decomissioning

As (3.1) is faster than (3.2) - the ingress basically makes the pending decomissiong broker unavailable for an external Kafka client.

**Describe the solution you'd like to see**

The operator also maintains the `status` sub-resource on KafkaCluster.
This is a more reliable source to drive the ingress configuration as it tracks all states for a kafka broker.
The envoy reconciler should be updated to watch this Status field and update the configuration accordingly.
 

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

**Comment by amuraru** [comment]
@baluchicken I believe istio ingress suffers from the same race on down scale - but didn't test that thoroughly.

**Comment by adamantal** [comment]
Hi @amuraru, @baluchicken, 

I found a regression related to this issue.
The fix works well when you downscale a cluster, but currently when you create a cluster with external listener config reconcile fails with this message:
```
2020-11-18T11:29:48.895Z    ERROR    controller    Reconciler error    {"reconcilerGroup": "kafka.banzaicloud.io", "reconcilerKind": "KafkaCluster", "controller": "KafkaCluster", "name": "kafka", "namespace": "kafka", "error": "creating resource failed: Service \"envoy-loadbalancer\" is invalid: spec.ports: Required value"}
```

The root cause is that currently the input for the envoy config is only the broker status slice, which is empty in case of a cold start (the brokers are not yet come alive).

Minimal reproduction of the issue:
1. Install the operator with dependencies
1. `kubectl create -f config/samples/kafkacluster_with_external_ssl.yaml`
1. Reconcile fails

I created a fix for this in #501, I'd be glad if you could take a look at it.
